### ISSUE TITLE
SingleRequest.java blocking main thread, ANR (w/ ThumbnailRequestCoordinator)

### ISSUE THREAD

**Issue by bwalsh0** [bumptech/glide#3667]
**Glide Version**: 4.9.0

**Integration libraries**: OkHttp3 [3.9.0]

**Device/Android Version**: 35+ different device models, all on Android 9.0 Pie

**Issue details / Repro steps / Use case background**: Downloads an image via URL into a `Bitmap` object and assigns it to a notification's thumbnail. Source images range from 500KB to 3MB.

I've done extensive testing with LeakCanary/different network, image size, and memory scenarios and have been unable to reproduce the ANR as reported in Play Console, with ~60 affected users. The ANR trace appears to point out some kind of timeout with the download and the maximum Android service time limit, but unsure if this is a Glide issue or improper use of it on my part. OkHttp3 timeouts are set to 90 seconds R/W/Connect. All affected users are on Android 9.0, possibly related to the SSL/cleartext restrictions in Pie.

**Glide load line / `GlideModule` (if any) / list Adapter code (if any)**:
Below code comes from: `BroadcastReciever` class -> `enqueueWork()` -> `JobIntentService` -> `AsyncTask` (below)
```java
   // This is inside doInBackground() of an AsyncTask, within the JobIntentService class.
                Bitmap notifThumb;
                try {
                    notifThumb = GlideApp.with(aContext.get()) //WeakReference.get to AppContext
                            .asBitmap()
                            .load(params[0]) //img direct url
                            .thumbnail(0.75f)
                            .override(1024, 512) //crop to thumbnail window size
                            .apply(new RequestOptions().skipMemoryCache(true).diskCacheStrategy(DiskCacheStrategy.NONE))
                            .into(1024,512) //possibly redundant
                            .get();
                } catch (ExecutionException | InterruptedException e) {
                    Crashlytics.setString("ThumbUrl", srcDirect);
                    Crashlytics.logException(e);
                    e.printStackTrace();
                }
```

No Layout XML involved. Downloads an image via URL into a `Bitmap` object and assigns it to a notification's thumbnail.

**Stack trace / LogCat**:
```ruby
"main" prio=5 tid=1 Blocked
  | group="main" sCount=1 dsCount=0 flags=1 obj=0x75352a18 self=0x76b5e14c00
  | sysTid=13298 nice=-4 cgrp=default sched=0/0 handle=0x773b8b1548
  | state=S schedstat=( 5520467430 3327803911 14046 ) utm=389 stm=162 core=3 HZ=100
  | stack=0x7fe9fc3000-0x7fe9fc5000 stackSize=8MB
  | held mutexes=
  at com.bumptech.glide.request.SingleRequest.isComplete (SingleRequest.java)
- waiting to lock <0x010c3574> (a com.bumptech.glide.request.SingleRequest) held by thread 62
  at com.bumptech.glide.request.ThumbnailRequestCoordinator.isComplete (ThumbnailRequestCoordinator.java:143)
  at com.bumptech.glide.manager.RequestTracker.restartRequests (RequestTracker.java:153)
  at com.bumptech.glide.RequestManager$RequestManagerConnectivityListener.onConnectivityChanged (RequestManager.java:675)
- locked <0x06344a9d> (a com.myappname.xxx.GlideRequests)
  at com.bumptech.glide.manager.DefaultConnectivityMonitor$1.onReceive (DefaultConnectivityMonitor.java:36)
  at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0 (LoadedApk.java:1391)
  at android.app.-$$Lambda$LoadedApk$ReceiverDispatcher$Args$_BumDX2UKsnxLVrE6UJsJZkotuA.run (lambda)
  at android.os.Handler.handleCallback (Handler.java:873)
  at android.os.Handler.dispatchMessage (Handler.java:99)
  at android.os.Looper.loop (Looper.java:193)
  at android.app.ActivityThread.main (ActivityThread.java:6718)
  at java.lang.reflect.Method.invoke (Method.java)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:493)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:858)
```

**Comment by sjudd** [comment]
Do you have the stack trace for the other thread (tid=62)?

**Comment by bwalsh0** [comment]
For this specific ANR report, thread 62 is not listed unfortunately, most of the traces stop around threads 30-50. I checked another report and found an very similar stack that did log a Glide thread.

Main thread says "held by thread 43", and here is 43 which is also blocked:

```ruby
"glide-source-thread-1" prio=5 tid=43 Blocked
  | group="main" sCount=1 dsCount=0 flags=1 obj=0x12d01f18 self=0x762cabf800
  | sysTid=32074 nice=9 cgrp=default sched=0/0 handle=0x7650cbd4f0
  | state=S schedstat=( 5979636 11002971 42 ) utm=0 stm=0 core=2 HZ=100
  | stack=0x7650bba000-0x7650bbc000 stackSize=1041KB
  | held mutexes=
  at com.bumptech.glide.request.SingleRequest.isResourceSet (SourceFile)
- waiting to lock <0x0b821f0b> (a com.bumptech.glide.request.SingleRequest) held by thread 14
  at com.bumptech.glide.request.ThumbnailRequestCoordinator.isResourceSet (SourceFile:1)
  at com.bumptech.glide.request.ThumbnailRequestCoordinator.isAnyResourceSet (SourceFile:1)
  at com.bumptech.glide.request.SingleRequest.i (SourceFile:1)
  at com.bumptech.glide.request.SingleRequest.a (SourceFile:62)
- locked <0x07f50dcc> (a com.bumptech.glide.request.SingleRequest)
  at com.bumptech.glide.request.SingleRequest.onLoadFailed (SourceFile:1)
- locked <0x07f50dcc> (a com.bumptech.glide.request.SingleRequest)
  at com.bumptech.glide.load.engine.h.a (SourceFile:17)
- locked <0x095ec1e8> (a com.bumptech.glide.load.engine.h)
  at com.bumptech.glide.load.engine.h$a.run (SourceFile:3)
- locked <0x095ec1e8> (a com.bumptech.glide.load.engine.h)
  at com.bumptech.glide.util.Executors$2.execute (SourceFile:1)
  at com.bumptech.glide.load.engine.h.c (SourceFile:23)
  at com.bumptech.glide.load.engine.h.onLoadFailed (SourceFile:4)
  at com.bumptech.glide.load.engine.g.e (SourceFile:3)
  at com.bumptech.glide.load.engine.g.i (SourceFile:10)
  at com.bumptech.glide.load.engine.g.j (SourceFile:4)
  at com.bumptech.glide.load.engine.g.run (SourceFile:7)
  at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
  at java.lang.Thread.run (Thread.java:764)
  at com.bumptech.glide.load.engine.executor.GlideExecutor$DefaultThreadFactory$1.run (SourceFile:8)
```

Unfortunately thread 14 is not logged, list goes from 13 and jumps to 15.

**Comment by sjudd** [comment]
It's still possible that there's deadlock, but I don't see it from these two stack traces. To tell for sure you'd really need the traces of all of the relevant threads from a single report.

Android will also report ANRs for broadcast receivers that go async and fail to stop in time, and may do the same for job scheduler services that take too long as well. I wouldn't expect to see the main thread stuck waiting for a lock though.

**Comment by fenjuly** [comment]
I have faced the similar block bug, I'am runing on 4.10.0-SNAPSHOT。
any idea of that?

```DefaultDispatcher-worker-11" daemon prio=5 tid=138 Blocked
  | group="main" sCount=1 dsCount=0 flags=1 obj=0x12c4eaa8 self=0xc82f8c00
  | sysTid=23268 nice=0 cgrp=default sched=0/0 handle=0xca93b970
  | state=S schedstat=( 158854486 148320521 761 ) utm=13 stm=2 core=5 HZ=100
  | stack=0xca838000-0xca83a000 stackSize=1042KB
  | held mutexes=
  at com.bumptech.glide.load.engine.Engine.load(Engine.java:189)
  - waiting to lock <0x03425ec5> (a com.bumptech.glide.load.engine.Engine) held by thread 1
  at com.bumptech.glide.request.SingleRequest.onSizeReady(SingleRequest.java:434)
  - locked <0x0e4b2e82> (a com.bumptech.glide.request.SingleRequest)
  at com.bumptech.glide.request.RequestFutureTarget.getSize(RequestFutureTarget.java:121)
  at com.bumptech.glide.request.SingleRequest.begin(SingleRequest.java:260)
  - locked <0x0e4b2e82> (a com.bumptech.glide.request.SingleRequest)
  at com.bumptech.glide.manager.RequestTracker.runRequest(RequestTracker.java:43)
  at com.bumptech.glide.RequestManager.track(RequestManager.java:645)
  - locked <0x037c1b09> (a com.sundayfun.daycam.GlideRequests)
  at com.bumptech.glide.RequestBuilder.into(RequestBuilder.java:642)
  at com.bumptech.glide.RequestBuilder.into(RequestBuilder.java:608)
  at com.bumptech.glide.RequestBuilder.submit(RequestBuilder.java:755)
  at com.bumptech.glide.RequestBuilder.submit(RequestBuilder.java:739)
  at com.sundayfun.daycam.camera.extension.MaskMetaExtension$loadMaskMeta$2$1$deferred$1.invokeSuspend(MaskMetaExtension.kt:510)
  at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:32)
  at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:233)
  at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
  at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742)```

**Comment by Raynor999** [comment]
### same issue, deadlock @sjudd 
### trace.txt,  main log infomation  
``` 
main" prio=5 tid=1 Blocked
  | group="main" sCount=1 dsCount=0 flags=1 obj=0x783c3d90 self=0xf2048000
  | sysTid=30349 nice=-10 cgrp=default sched=0/0 handle=0xf5f3a494
  | state=S schedstat=( 31318042518 7263808444 51603 ) utm=2690 stm=440 core=7 HZ=100
  | stack=0xff516000-0xff518000 stackSize=8MB
  | held mutexes=
  at com.bumptech.glide.RequestManager.clear(RequestManager.java:-1)
  - waiting to lock <0x006dc09e> (a com.sundayfun.daycam.GlideRequests) held by thread 153
  at com.bumptech.glide.load.resource.gif.GifFrameLoader.clear(GifFrameLoader.java:187)
  at com.bumptech.glide.load.resource.gif.GifDrawable.recycle(GifDrawable.java:385)
  at com.bumptech.glide.load.resource.gif.GifDrawableResource.recycle(GifDrawableResource.java:32)
  at com.bumptech.glide.load.engine.EngineResource.recycle(EngineResource.java:73)
  - locked <0x079f577f> (a com.bumptech.glide.load.engine.EngineResource)
  at com.bumptech.glide.load.engine.ResourceRecycler.recycle(ResourceRecycler.java:24)
  - locked <0x06bb234c> (a com.bumptech.glide.load.engine.ResourceRecycler)
  at com.bumptech.glide.load.engine.Engine.onResourceRemoved(Engine.java:315)
  at com.bumptech.glide.load.engine.cache.LruResourceCache.onItemEvicted(LruResourceCache.java:33)
  at com.bumptech.glide.load.engine.cache.LruResourceCache.onItemEvicted(LruResourceCache.java:13)
  at com.bumptech.glide.util.LruCache.trimToSize(LruCache.java:183)
  - locked <0x0b375f95> (a com.bumptech.glide.load.engine.cache.LruResourceCache)
  at com.bumptech.glide.util.LruCache.evict(LruCache.java:188)
  at com.bumptech.glide.util.LruCache.put(LruCache.java:141)
  - locked <0x0b375f95> (a com.bumptech.glide.load.engine.cache.LruResourceCache)
  at com.bumptech.glide.load.engine.cache.LruResourceCache.put(LruResourceCache.java:13)
  at com.bumptech.glide.load.engine.Engine.onResourceReleased(Engine.java:322)
  - locked <0x07f148aa> (a com.bumptech.glide.load.engine.Engine)
  at com.bumptech.glide.load.engine.EngineResource.release(EngineResource.java:113)
  - locked <0x07f148aa> (a com.bumptech.glide.load.engine.Engine)
  - locked <0x035bec9b> (a com.bumptech.glide.load.engine.EngineResource)
  at com.bumptech.glide.load.engine.Engine.release(Engine.java:286)
  at com.bumptech.glide.request.SingleRequest.releaseResource(SingleRequest.java:340)
  at com.bumptech.glide.request.SingleRequest.clear(SingleRequest.java:330)
  - locked <0x0791b9bc> (a com.bumptech.glide.request.SingleRequest)
  at com.bumptech.glide.manager.RequestTracker.clearRemoveAndMaybeRecycle(RequestTracker.java:79)
  at com.bumptech.glide.manager.RequestTracker.clearRemoveAndRecycle(RequestTracker.java:66)
  at com.bumptech.glide.RequestManager.untrack(RequestManager.java:630)
  - locked <0x0cd6751c> (a com.sundayfun.daycam.GlideRequests)
  at com.bumptech.glide.RequestManager.untrackOrDelegate(RequestManager.java:598)
  at com.bumptech.glide.RequestManager.clear(RequestManager.java:594)
  - locked <0x0cd6751c> (a com.sundayfun.daycam.GlideRequests)
  at com.bumptech.glide.RequestManager.onDestroy(RequestManager.java:339)
  - locked <0x0cd6751c> (a com.sundayfun.daycam.GlideRequests)
  at com.bumptech.glide.manager.ActivityFragmentLifecycle.onDestroy(ActivityFragmentLifecycle.java:65)
  at com.bumptech.glide.manager.SupportRequestManagerFragment.onDestroy(SupportRequestManagerFragment.java:194)
  at androidx.fragment.app.Fragment.performDestroy(Fragment.java:2800)`
===== **Thread 153**
`"DefaultDispatcher-worker-7" daemon prio=5 tid=153 Blocked
  | group="main" sCount=1 dsCount=0 flags=1 obj=0x174029f8 self=0xb68e6e00
  | sysTid=32510 nice=0 cgrp=default sched=0/0 handle=0xb90e3970
  | state=S schedstat=( 413739126 668182464 3037 ) utm=24 stm=16 core=6 HZ=100
  | stack=0xb8fe0000-0xb8fe2000 stackSize=1042KB
  | held mutexes=
  at com.bumptech.glide.load.engine.Engine.load(Engine.java:-1)
  - waiting to lock <0x07f148aa> (a com.bumptech.glide.load.engine.Engine) held by thread 1
  at com.bumptech.glide.request.SingleRequest.onSizeReady(SingleRequest.java:447)
  - locked <0x044650d1> (a com.bumptech.glide.request.SingleRequest)
  at com.bumptech.glide.request.RequestFutureTarget.getSize(RequestFutureTarget.java:125)
  at com.bumptech.glide.request.SingleRequest.begin(SingleRequest.java:272)
  - locked <0x044650d1> (a com.bumptech.glide.request.SingleRequest)
  at com.bumptech.glide.manager.RequestTracker.runRequest(RequestTracker.java:44)
  at com.bumptech.glide.RequestManager.track(RequestManager.java:641)
  - locked <0x006dc09e> (a com.sundayfun.daycam.GlideRequests)
  at com.bumptech.glide.RequestBuilder.into(RequestBuilder.java:656)
  at com.bumptech.glide.RequestBuilder.into(RequestBuilder.java:622)
  at com.bumptech.glide.RequestBuilder.submit(RequestBuilder.java:775)
  at com.bumptech.glide.RequestBuilder.submit(RequestBuilder.java:757)
  at com.sundayfun.daycam.camera.extension.MaskMetaExtension$loadMaskMeta$2$1$deferred$1.invokeSuspend(MaskMetaExtension.kt:510)
  at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:32)
  at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:233)
  at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
  at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742)`

It looks like the main thread is waiting for a XXXRequestManger lock, but the lock is held by a child thread that is waiting for the main thread to release the Engine lock.

But after analyzing the source code and trace log, I think the child thread holds the "ApplicationRequestManger" object lock， and the main thread holds the "RequestManger" lock in the SupportRequestManagerFragment。So, Should not be deadlocked?

**Comment by sjudd** [comment]
@Raynor999 your stack traces look like deadlock, but they're not similar to the original stack traces in this issue. Let's follow up in #3684 

I believe the internal version of the original issue is b/138649092, where I have a complete set of stack traces.

**Comment by amiton** [comment]
@sjudd Thanks for all the amazing work. We are also experiencing this deadlock issue causing ANR's and i saw that you committed a fix. When would you expect a release of Glide version 4.10.0 including this fix?

**Comment by amiton** [comment]
@sjudd Any estimate on when to expect the release of Glide 4.10.0?

**Comment by sjudd** [comment]
This should be fixed by b2a46ef7a385abd5df38ee11b8be85868768739e. 

I'm hoping to get a release out soon. I haven't seen new deadlocks with these changes. Work has been a bit busier than I'd like the last couple weeks.
